File: tests/07_phases_ompss.dg/fortran/success_taskloop_06.f90
! <testinfo>
! test_generator=config/mercurium-ompss
! </testinfo>
PROGRAM P
    IMPLICIT NONE
    INTEGER :: G, I
    INTEGER, PARAMETER :: MAX_NUM_TASKS = 20
    INTEGER, PARAMETER :: N = 2000
    INTEGER :: A(N)

    INTEGER :: L, U, S

    
    A = 0
    L = 1
    U = 2000
    S = 1
    DO G=1, MAX_NUM_TASKS

        !$OMP TASKLOOP SHARED(A) NUM_TASKS(G)
        DO I=L, U, S
            A(I) = A(I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_NUM_TASKS)) STOP -1


    A = 0
    L = 1
    U = 2000
    S = 3
    DO G=1, MAX_NUM_TASKS

        !$OMP TASKLOOP SHARED(A) NUM_TASKS(G)
        DO I=L, U, S
            A(I) = A(I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_NUM_TASKS)) STOP -2


    A = 0
    L = 2000
    U = 1
    S = -1
    DO G=1, MAX_NUM_TASKS

        !$OMP TASKLOOP SHARED(A) NUM_TASKS(G)
        DO I=L, U, S
            A(I) = A(I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_NUM_TASKS)) STOP -3


    A = 0
    L = 2000
    U = 1
    S = -3
    DO G=1, MAX_NUM_TASKS

        !$OMP TASKLOOP SHARED(A) NUM_TASKS(G)
        DO I=L, U, S
            A(I) = A(I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_NUM_TASKS)) STOP -4


    A = 0
    L = -999
    U = 1000
    S = 1
    DO G=1, MAX_NUM_TASKS

        !$OMP TASKLOOP SHARED(A) NUM_TASKS(G)
        DO I=L, U, S
            A(I+1000) = A(I+1000) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_NUM_TASKS)) STOP -5


    A = 0
    L = -999
    U = 1000
    S = 3
    DO G=1, MAX_NUM_TASKS

        !$OMP TASKLOOP SHARED(A) NUM_TASKS(G)
        DO I=L, U, S
            A(I+1000) = A(I+1000) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_NUM_TASKS)) STOP -6


    A = 0
    L = 999
    U = -1000
    S = -1
    DO G=1, MAX_NUM_TASKS

        !$OMP TASKLOOP SHARED(A) NUM_TASKS(G)
        DO I=L, U, S
            A(1000 - I) = A(1000 - I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_NUM_TASKS)) STOP -7


    A = 0
    L = 999
    U = -1000
    S = -3
    DO G=1, MAX_NUM_TASKS

        !$OMP TASKLOOP SHARED(A) NUM_TASKS(G)
        DO I=L, U, S
            A(1000 - I) = A(1000 - I) + 1
        ENDDO
    ENDDO
    IF(ANY(A(::S) /= MAX_NUM_TASKS)) STOP -8

END PROGRAM P

File: tests/07_phases_ompss.dg/cxx/success_taskloop_06.cpp
/*--------------------------------------------------------------------
  (C) Copyright 2006-2012 Barcelona Supercomputing Center
                          Centro Nacional de Supercomputacion
  
  This file is part of Mercurium C/C++ source-to-source compiler.
  
  See AUTHORS file in the top level directory for information
  regarding developers and contributors.
  
  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 3 of the License, or (at your option) any later version.
  
  Mercurium C/C++ source-to-source compiler is distributed in the hope
  that it will be useful, but WITHOUT ANY WARRANTY; without even the
  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
  PURPOSE.  See the GNU Lesser General Public License for more
  details.
  
  You should have received a copy of the GNU Lesser General Public
  License along with Mercurium C/C++ source-to-source compiler; if
  not, write to the Free Software Foundation, Inc., 675 Mass Ave,
  Cambridge, MA 02139, USA.
--------------------------------------------------------------------*/



/*
<testinfo>
test_generator=config/mercurium-ompss
</testinfo>
*/

#include <assert.h>

void loop_1(int *v, int start, int end, int step, int offset, int chunk)
{
    #pragma omp taskloop num_tasks(chunk)
    for (int i = start; i < end; i += step) {
        v[i+offset]++;
    }
}

void loop_2(int *v, int start, int end, int step, int offset, int chunk)
{
    #pragma omp taskloop num_tasks(chunk)
    for (int i = start; i > end; i += step) {
        v[offset-i]++;
    }
}


void check_results(int *v, int start, int end, int step, int chunk_values)
{
    for (int i = start; i < end; i += step)
        assert(v[i] == chunk_values);
}


void init(int *v, int size)
{
    for(int i = 0; i < size; ++i)
        v[i] = 0;
}

int main(int argc, char *argv[]) {
    const int n = 2000;
    int v[n];


    //printf("first config\n");
    {
        int start  =     0;
        int end    =  2000;
        int step   =    10;
        int offset =     0;

        int chunk_values = 20;
        init(v, n);
        for (int i = 1; i <= chunk_values; i++)
            loop_1(v, start, end, step, offset, i);

        check_results(v, 0, n, step, 20);
    }

    //printf("second config\n");
    {
        int start  =  2000;
        int end    =     0;
        int step   =   -10;
        int offset =  2000;

        int chunk_values = 20;
        init(v, n);
        for (int i = 1; i <= chunk_values; i++)
            loop_2(v, start, end, step, offset, i);

        check_results(v, 0, n, -step, 20);
    }

    //printf("third config\n");
    {
        int start  = -2000;
        int end    =     0;
        int step   =    10;
        int offset =  2000;

        int chunk_values = 20;
        init(v, n);
        for (int i = 1; i <= chunk_values; i++)
            loop_1(v, start, end, step, offset, i);

        check_results(v, 0, n, step, 20);
    }


    //printf("fourth config\n");
    {
        int start  =     0;
        int end    = -2000;
        int step   =   -10;
        int offset =     0;

        int chunk_values = 20;
        init(v, n);
        for (int i = 1; i <= chunk_values; i++)
            loop_2(v, start, end, step, offset, i);

        check_results(v, 0, n, -step, 20);
    }


    //printf("fifth config\n");
    {
        int start  = -1000;
        int end    =  1000;
        int step   =    10;
        int offset =  1000;

        int chunk_values = 20;
        init(v, n);
        for (int i = 1; i <= chunk_values; i++)
            loop_1(v, start, end, step, offset, i);

        check_results(v, 0, n, step, 20);
    }

    //printf("sixth config\n");
    {
        int start  =  1000;
        int end    = -1000;
        int step   =   -10;
        int offset =  1000;

        int chunk_values = 20;
        init(v, n);
        for (int i = 1; i <= chunk_values; i++)
            loop_2(v, start, end, step, offset, 10);

        check_results(v, 0, n, -step, chunk_values);
    }

    return 0;
}
